aws-s3-transfer-manager/src/operation/download/discovery.rs

```diff
@@ -181,10 +136,107 @@ async fn discover_obj_with_get(
     })
 }
 
+async fn discover_obj_with_get_first_part(
+    ctx: &DownloadContext,
+    input: &DownloadInput,
+) -> Result<ObjectDiscovery, error::Error> {
+    // Get object first part.
+    let builder: GetObjectInputBuilder = input.clone().into();
+    // S3 index starts with 1
+    let resp = builder
+        .set_range(None)
+        .set_part_number(Some(1))
+        .send_with(ctx.client())
+        .await
+        .map_err(error::discovery_failed)?;
+    discover_handler_first_chunk_response(resp, None)
+}
+
+async fn discover_obj_with_head(
+    ctx: &DownloadContext,
+    input: &DownloadInput,
+    byte_range: Option<ByteRange>,
+) -> Result<ObjectDiscovery, crate::error::Error> {
+    let resp = ctx
+        .client()
+        .head_object()
+        .set_bucket(input.bucket().map(str::to_string))
+        .set_key(input.key().map(str::to_string))
+        .send()
+        .await;
+
+    match resp {
+        Err(error) => {
+            match error.as_service_error() {
+                Some(service_error) if service_error.meta().code() == Some("InvalidRange") => {
+                    // Invalid Range because of the empty object, discover the object with the first part instead.
+                    discover_obj_with_get_first_part(ctx, input).await
+                }
+                _ => Err(error::discovery_failed(error)),
+            }
+        }
+        Ok(resp) => {
+            let object_meta: ObjectMetadata = resp.into();
+
+            let remaining = match byte_range {
+                Some(range) => match range {
+                    ByteRange::Inclusive(start, end) => start..=end,
+                    ByteRange::AllFrom(start) => start..=object_meta.content_length(),
+                    ByteRange::Last(n) => {
+                        (object_meta.content_length() - n + 1)..=object_meta.content_length()
+                    }
+                },
+                None => 0..=object_meta.content_length(),
+            };
+
+            Ok(ObjectDiscovery {
+                remaining,
+                chunk_meta: None,
+                object_meta,
+                initial_chunk: None,
+            })
+        }
+    }
+}
+
+async fn discover_obj_with_get(
+    ctx: &DownloadContext,
+    input: &DownloadInput,
+    range: Option<RangeInclusive<u64>>,
```

Comment:
however we rename `range` in that other function, rename it the same way here too